File: May-18/28-05-18/python/01-Dhoom-4.py
N = 10**5 + 5
visit = [-1]*N
 
def solve(ar, n, k, l):
    q = []
    q.append(k)
    visit[k] = 0
    
    
    while(len(q) > 0):
        val = q[0]
        del q[0]
        
        if(val == l):
            print(visit[l])
            return
        
        for i in range(0,n):
            data = (val * ar[i]) % 100000
            if(visit[data] == -1):
                visit[data] = 1 + visit[val]
                q.append(data)
            
            if(data == l):
                print(visit[l])
                return
                
    print(visit[l])
    
k, l = map(int, input().split())
n = int(input())
ar = [int(x) for x in input().split()]
solve(ar, n, k, l)

File: May-18/28-05-18/java/01-Dhoom-4.java

import java.util.*;


class TestClass {
    static int N = (int)1e5 + 5;
    static int visit[] = new int[N];
    
    static void solu(long ar[], long n, long k, long l)
    {
        for(int i = 0; i < N; i++)
            visit[i] = -1;
        
        LinkedList<Long> q = new LinkedList<Long>();
        q.add(k);
        visit[(int)k] = 0;
        
        while(q.size() != 0)
        {
            long val = q.poll();
            
            for(int i = 0; i < n; i++)
            {
                long data = (val * ar[i]) % 100000;
                if(visit[(int)data] == -1)
                {
                    visit[(int)data] = 1 + visit[(int)val];
                    q.add(data);
                }
            }
        }
        System.out.println(visit[(int)l]);
        
    }
    public static void main(String args[] ) throws Exception {
        Scanner input = new Scanner(System.in);
        long k, l;
        k = input.nextLong();
        l = input.nextLong();
        long n = input.nextLong();
        long ar[] = new long[(int)n];
        for(int i = 0; i < n; i++)
            ar[i] = input.nextLong();
        solu(ar, n, k, l);
    }
}
